### PR TITLE
Oauth2

### PR DESCRIPTION
**Pull request by AndyWendt**

### CODE REVIEW

**Comment by AndyWendt** [comment]
This is a WIP and not mergeable until stated otherwise

File: lib/xeroizer/generic_application.rb [hunk 1]
```diff
@@ -63,7 +63,7 @@ class GenericApplication
       # @see PublicApplication
       # @see PrivateApplication
       # @see PartnerApplication
-      def initialize(consumer_key, consumer_secret, options = {})
+      def initialize(client, options = {})
```
Comment (CloCkWeRX) [inline lib/xeroizer/generic_application.rb]: With this, may be worth doing a major version bump, so people are aware of the constructor change.

File: lib/xeroizer/generic_application.rb [hunk 2]
```diff
@@ -72,7 +72,7 @@ def initialize(consumer_key, consumer_secret, options = {})
         @before_request = options.delete(:before_request)
         @after_request = options.delete(:after_request)
         @around_request = options.delete(:around_request)
-        @client = OAuth.new(consumer_key, consumer_secret, options.merge({default_headers: default_headers}))
+        @client = client
```
Comment (CloCkWeRX) [inline lib/xeroizer/generic_application.rb]: Consider checking client.kind_if?(OAuth) or similar; raising an argument error if given strings accidentally.

File: lib/xeroizer/oauth2_factory.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+module Xeroizer
+   class OAuth2Factory
```
Comment (CloCkWeRX) [inline lib/xeroizer/oauth2_factory.rb]: Maybe skip the factory pattern, unless there's a lot of wiring up that happens?

Its slightly more verbose to do 
```Xeroizer::OAuth2Factory.build(x, y, z)```
than
```Xeroizer::OAuth2.new(x, y, z)```

If it had some logic to it, I could see the use.

File: test/unit_test_helper.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
+require 'test_helper'
+require 'webmock'
+
+include WebMock::API
+WebMock.disable_net_connect!(allow_localhost: true)
```
Comment (CloCkWeRX) [inline test/unit_test_helper.rb]: May be worth pulling this in as its own PR; possibly using VCR or similar for easy recording and replay of cassettes.

**Comment by CloCkWeRX** [comment]
Looks great so far! Thanks for taking this on.

**Comment by daande** [comment]
@AndyWendt Is it possible to get an ETA on this release? Or have this released soon?

File: lib/xeroizer/generic_application.rb (hunk 1 above)
Comment (AndyWendt) [reply lib/xeroizer/generic_application.rb]: The consumers likely shouldn't see any change, since we didn't change the other clients.  However, if they depend on the generic application then there would be a problem.  

We've strived to maintain passing tests in the unit and integration/acceptance suite.

**Comment by AndyWendt** [comment]
@daande we're using this for our own purposes right now.  If you need changes to what we've done, let us know.  

We're going to be revising this once we deploy for a client commitment.

File: test/unit_test_helper.rb (hunk 4 above)
Comment (AndyWendt) [reply test/unit_test_helper.rb]: That would be a good idea.

File: test/acceptance/acceptance_test.rb [hunk 5]
```diff
@@ -2,6 +2,11 @@ module AcceptanceTest
   class << self
     def included(klass)
       klass.class_eval do
+        def self.it_works_using_oauth1_and_oauth2(&block)
```
[the remark this answers is not in the document]
Comment (AndyWendt) [reply test/acceptance/acceptance_test.rb]: I expect you're right on this.  This is some of the code that we need to unwind.

File: lib/xeroizer/oauth2_factory.rb (hunk 3 above)
Comment (AndyWendt) [reply lib/xeroizer/oauth2_factory.rb]: Some more code we need to unwind

File: lib/xeroizer/generic_application.rb (hunk 2 above)
Comment (AndyWendt) [reply lib/xeroizer/generic_application.rb]: We can do that.

**Comment by daande** [comment]
> @daande we're using this for our own purposes right now. If you need changes to what we've done, let us know.
> 
> We're going to be revising this once we deploy for a client commitment. I expect a releasable version won't be available for a few to several weeks because of another system that we're working on that uses this code.
> 
> Once that's done, we have bandwidth to tidy this up. It works but not ready for general release.

Thanks for the update!

**Comment by aovertus** [comment]
@AndyWendt is this "workable"? I've to upgrade our project to use Xero with oauth2, we're currently using Xeroizer with oauth1.

**Comment by joegaudet** [comment]
Cross-posted:

Hey all, great work on the gem, I was wondering in the short term, if we manage the oath loop our selves if there's a hack to have the gem use the token we manually manage. I've been experimenting with the official gem, but it doesn't feel ready for prime time yet.

**Comment by AndyWendt** [comment]
The gem is workable, however, you have to manage the tokens yourself.

**Comment by joegaudet** [comment]
Would we need to do that with this branch version? Or could we use the
released one? I've already got token management working on my own.

File: lib/xeroizer/generic_application.rb (hunk 1 above)
Comment (AndyWendt) [reply lib/xeroizer/generic_application.rb]: @CloCkWeRX I'll let you make the determination as to whether or not you want to make a major version bump here.

File: lib/xeroizer/generic_application.rb (hunk 2 above)
Comment (AndyWendt) [reply lib/xeroizer/generic_application.rb]: Resolved above

File: test/acceptance/about_creating_bank_transactions_test.rb [hunk 6]
```diff
@@ -4,40 +4,67 @@
...
-    Xeroizer::PrivateApplication.new(@consumer_key, @consumer_secret, @key_file)
+  def assert_exists(bank_transaction, client)
+    assert_not_nil bank_transaction.id,
+                   "Cannot check for exitence unless the bank transaction has non-null identifier"
+    assert_not_nil client.BankTransaction.find bank_transaction.id
+  end
+
+  def any_line_items(account)
+    [{
+         :description => "Clingfilm bike shorts",
```
Comment (CloCkWeRX) [inline test/acceptance/about_creating_bank_transactions_test.rb]: Prefer modern syntax (rubocop -a should fix it for you)

File: lib/xeroizer/generic_application.rb (hunk 1 above)
Comment (CloCkWeRX) [reply lib/xeroizer/generic_application.rb]: I'm happy to skip it then

**Comment by CloCkWeRX** [comment]
Thanks for your efforts here!

**Comment by nahzz** [comment]
@AndyWendt any chance to make a new gem release for the OAuth2 support ?  Thanks

**Comment by pxlpnk** [comment]
@AndyWendt Would really apprechiate a new release as support for OAuth1 is ceased and we can't even generate new credentials for this version anymore.

**Comment by AndyWendt** [comment]
@pxlpnk @nahzz I am not a maintainer and I don't have access to release a new version.

**Comment by SidneyAllen** [comment]
@AndyWendt @pxlpnk and @nahzz 

Wayne gave us access to build and release new version of xeroizer, but we'll need to rely on your folks to confirm the code in master is ready for release and the recommended versioning. (patch, min, major release) Would this be 3.0.0 ?

**Comment by nahzz** [comment]
@SidneyAllen I made some tests myself and looks like it's working at least for contacts and bills .  But it seems it doesn't raise the errors such as TokenExpired anymore (I deliberately used an expired token ,  and set raise_errors to true when constructing a OAuth2Application).

@AndyWendt  Any ideas to raise the errors such as "RateLimitExceeded"  or "TokenExpired" for oauth2 ,  like in the OAuth1 ?

**Comment by AndyWendt** [comment]
Sorry these update emails keep going to spam.  @SidneyAllen @nahzz 

I can look into the errors as part of my contributing back portion of my week.

**Comment by nahzz** [comment]
@AndyWendt  thanks !  btw:  is there a type[ here ](https://github.com/waynerobinson/xeroizer/blob/master/lib/xeroizer/oauth2_application.rb#L22)? (tenents_url should be tenants_url ? )

**Comment by AndyWendt** [comment]
@nahzz yeah... that's a typo

**Comment by nahzz** [comment]
@AndyWendt  does it affect the functionalities of the code in any way ?  looks like it doesn't

**Comment by AndyWendt** [comment]
@nahzz I don't believe so

**Comment by LimeBlast** [comment]
Is someone able to provide a list of all the [scopes](https://developer.xero.com/documentation/oauth2/scopes) which will be required by the Xeroizer gem?

So far I've figured out that the following are required at a minimum, but wasn't sure which the others will be required.

```
openid profile email offline_access
```

Thanks.

**Comment by nahzz** [comment]
@LimeBlast I don't think `openid profile email` are required at a minimum, they are required if you want to access all the user details (email, last name, etc).  Also those three scopes won't be able to migrated from OAuth1 to OAuth2.

I also think scopes have nothing to do with Xeroizer gem.   It depends on what you wan to interact with Xero.  For example :  if you want to send `invoices` to Xero,  you would need to put `accounting.transactions` into your auth request to Xero (something like `https://login.xero.com/identity/connect/authorize?response_type=code&client_id=YOURCLIENTID&redirect_uri=YOURREDIRECTURI&scope=offline_access accounting.transactions&state=123`, if your client accept them and allow access, then you can use Xeroizer gem to send invoices to Xero

**Comment by LimeBlast** [comment]
My apologies, I've maybe misunderstood how Xeroizer works.

To provide some context, our company has inherited an app from another developer, an app which was in a sorry state, written by a developer that didn't believe in documentation.

Anyway, I've been tasked with updating Xeroizer to use oAuth 2.0. I know very little about what Xeroizer does, or how it's configured, so in an attempt to make my changes as low impact as possible, I updated it to use the `Xeroizer::OAuth2Application` initializer (as described above), and set about figuring out how to supply the required tokens.

To this end, I chose to use the [Xero gem](https://github.com/XeroAPI/xero-ruby) (inspired by the [https://github.com/XeroAPI/xero-ruby-oauth2-app](example app)), and have a working example of being able to connect to an oAuth 2.0 app, with refreshing tokens, etc.. - so all that remains is to ensure the scopes I need are set.

Now I think I was under the impression that Xeroizer had a set list of functions, for which I'd be able to get some scopes, but I think I now understand that it's just a wrapper around the Xero API? One which I could possibly do away with, now that I've set-up the Xero gem? (not that I plan on making any such changes).

So if I understand this correctly, it's not a case of the scopes which Xeroizer needs, but instead the scopes my app needs, based on the desired functionally within?

In any case, I've chosen the following scopes, which will hopefully provide everything I need.

```
openid profile email accounting.settings accounting.reports.read accounting.journals.read accounting.contacts accounting.attachments accounting.transactions assets assets.read projects projects.read offline_access
```

Thank you for your help, @nahzz

**Comment by nahzz** [comment]
@LimeBlast  Yeah Xeroizer is a wrapper around the Xero API ( sort of like the official Xero-ruby gem you mentioned),  it is the "official" ruby gem for Xero OAuth1.0, see here :  https://developer.xero.com/documentation/libraries/oauth1-sdks .  But from OAuth2 Xero starts to have their own gem,  which is the one you chose to use.

Indeed it's about the scope that your app needs, not what Xeroizer needs. 

For your scopes,  if I understand correctly probably `assets.read` is already included in `assets`.  From Xero's [site](https://developer.xero.com/documentation/oauth2/scopes)  `assets.read` is `As above but GET only`

**Comment by ZacHaque** [comment]
@LimeBlast I am a newbie to ruby and facing the same challenge as yours. Could you please advise the broken steps you followed to solve this problem. Are you ending up changing to xero-gem? it seems a lot of code change then?? if I want to stay with Xeroizer what needs to be done to use Oauth2

**Comment by LimeBlast** [comment]
Hi @ZacHaque 

I mostly just followed the example set by https://github.com/XeroAPI/xero-ruby-oauth2-starter for generating and refreshing access keys, which I then pass into Xeroizer each time I needed it. I'm not sure I can be much more help than that, however, as I'm not fully sure I understand it all myself.

**Comment by mrhillman** [comment]
Another potentially useful resource for refreshing tokens etc: https://github.com/XeroAPI/xero-ruby/blob/master/lib/xero-ruby/api_client.rb#L94

Likewise, I didn't want to switch away from Xeroizer, as it was all there (and xero-ruby had clashing dependencies which I didn't want to deal with)

While not fully built yet, I've tested through console and I injected my app with the initial access_token and refresh_token from the hacky xoauth thing then basically replicated the method in the link above to refresh when needed

**Comment by LimeBlast** [comment]
Thanks for that, @mrhillman - my implementation is still very rough around the edges, so I'll see if I can take advantage of that method. 👍

**Comment by nahzz** [comment]
you can also utilize https://github.com/oauth-xx/oauth2  gem to do those token related operations for you (get token ,refresh token, etc)
